impl/src/settings.rs

```diff
@@ -483,77 +468,299 @@ fn validate_settings(
+struct RawRazeSettings {
+  #[serde(default)]
+  pub workspace_path: Option<String>,
+  #[serde(default)]
+  pub package_aliases_dir: Option<String>,
+  #[serde(default)]
+  pub target: Option<String>,
+  #[serde(default)]
+  pub targets: Option<Vec<String>>,
+  #[serde(default)]
+  pub binary_deps: HashMap<String, cargo_toml::Dependency>,
+  #[serde(default)]
+  pub crates: HashMap<String, CrateSettingsPerVersion>,
+  #[serde(default)]
+  pub gen_workspace_prefix: Option<String>,
+  #[serde(default)]
+  pub genmode: Option<GenMode>,
+  #[serde(default)]
+  pub output_buildfile_suffix: Option<String>,
+  #[serde(default)]
+  pub default_gen_buildrs: Option<bool>,
+  #[serde(default)]
+  pub registry: Option<String>,
+  #[serde(default)]
+  pub index_url: Option<String>,
+  #[serde(default)]
+  pub incompatible_relative_workspace_path: Option<bool>,
+  #[serde(default)]
+  pub rust_rules_workspace_name: Option<String>,
+  #[serde(default)]
+  pub vendor_dir: Option<String>,
+}
+
+impl RawRazeSettings {
+  /** Checks whether or not the settings have non-package specific settings specified */
+  fn contains_primary_options(&self) -> bool {
+    self.workspace_path.is_some()
+      || self.package_aliases_dir.is_some()
+      || self.target.is_some()
+      || self.targets.is_some()
+      || self.gen_workspace_prefix.is_some()
+      || self.genmode.is_some()
+      || self.output_buildfile_suffix.is_some()
+      || self.default_gen_buildrs.is_some()
+      || self.registry.is_some()
+      || self.index_url.is_some()
+      || self.incompatible_relative_workspace_path.is_some()
+      || self.rust_rules_workspace_name.is_some()
+      || self.vendor_dir.is_some()
+  }
+}
+
+/** Grows a list with duplicate keys between two maps */
+fn extend_duplicates<K: Hash + Eq + Clone, V>(
+  extended_list: &mut Vec<K>,
+  main_map: &HashMap<K, V>,
+  input_map: &HashMap<K, V>,
+) {
+  extended_list.extend(
+    input_map
+      .iter()
+      .filter_map(|(key, _value)| {
+        // Log the key if it exists in both the main and input maps
+        if main_map.contains_key(key) {
+          Some(key)
+        } else {
+          None
+        }
+      })
+      .cloned()
+      .collect::<Vec<K>>(),
+  );
+}
+
+/** Parse [`RazeSettings`](crate::settings::RazeSettings) from workspace metadata */
+fn parse_raze_settings_workspace(
+  metadata_value: &serde_json::value::Value,
+  metadata: &Metadata,
+) -> Result<RazeSettings> {
+  let mut settings = RazeSettings::deserialize(metadata_value)?;
+
+  let workspace_packages: Vec<&Package> = metadata
+    .packages
+    .iter()
+    .filter(|pkg| metadata.workspace_members.contains(&pkg.id))
+    .collect();
+
+  let mut duplicate_binary_deps = Vec::new();
+  let mut duplicate_crate_settings = Vec::new();
+
+  for package in workspace_packages.iter() {
+    if let Some(pkg_value) = package.metadata.get("raze") {
+      let pkg_settings = RawRazeSettings::deserialize(pkg_value)?;
```

Comment:
Did you consider parsing workspace package settings using a different type? That seems like the more natural approach, I'm guessing you have a reason.

(I argue that's the more natural approach because you can then disallow "primary" fields by omitting them in the workspace package raze settings. You could argue that this results in two settings objects needing to be maintained in order, but the current approach requires additions to "primary" settings to be included in the "is_primary" fn, which I think is comparable.)

Reply:
I wasn't sure how best to do that so opted for this approach since I thought it was more straight forward. I feel there should only be 1 struct that contains all relevant settings/information that users interact with/care about. Having this be 1 struct also simplifies users ability to locate and understand settings available to them. In another change, I'd also like to add [deny_unknown_fields](https://serde.rs/container-attrs.html#deny_unknown_fields) to these structs to eliminate a sizable point of confusion which is called out in https://github.com/google/cargo-raze/issues/48. I agree having to manage two places for the same information is not ideal. Could you recommend something more structured here? Happy to make improvements 😄

Reply:
One other thing to consider, with this implementation I found it quite nice to be able to define deprecations: https://github.com/google/cargo-raze/pull/303/files#r542809410 whatever design this ends up being, I think there should be a clear, single place where deprecation warnings can be found.

Reply:
Hmm, alright.